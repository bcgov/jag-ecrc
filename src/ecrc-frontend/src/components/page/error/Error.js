import React, { useState } from "react";
import { Redirect } from "react-router-dom";
import PropTypes from "prop-types";

import "../page.css";
import Header from "../../base/header/Header";
import Footer from "../../base/footer/Footer";
import { Button } from "../../base/button/Button";

export default function Error({ page: { header, error } }) {
  const [toHome, setToHome] = useState(false);

  const homeButton = {
    label: "Home",
    buttonStyle: "btn ecrc_go_btn",
    buttonSize: "btn",
    type: "submit"
  };

  if (toHome) {
    return <Redirect to="/" />;
  }

  let errorContent;
  if (!error) {
    errorContent = (
      <div>
        <h1>Service is currently unavailable</h1>
        <p>Please try again later.</p>
      </div>
    );
  } else if (error.status === 403) {
    if (error.message === "BCSC login failed") {
      errorContent = (
        <div>
          <h1>Login failed</h1>
          <p>
            Unable to login using your BCSC account, please try again later.
          </p>
        </div>
      );
    } else if (error.message === "User is under the age of 12") {
      errorContent = (
        <div>
          <h1>Unauthorized age group</h1>
          <p>
            To submit an online request for a criminal record check, you must be
            at least 12 years of age.
          </p>
        </div>
      );
    } else {
      errorContent = (
        <div>
          <h1>Unauthorized entry</h1>
          <p>
            Unauthorized user entry, please return to the home page and begin
            your session again.
          </p>
        </div>
      );
    }
  } else if (error.status === 590) {
    errorContent = (
      <div>
        <h1>Your session has expired</h1>
        <p>Please return to the home page and begin your session again.</p>
      </div>
    );
  } else {
    errorContent = (
      <div>
        <h1>An unknown error has occurred</h1>
        <p>
          The error description is below. If this error persists, please try
          again later.
          <br />
          <br />
          {error.message}
        </p>
      </div>
    );
  }

  return (
    <main>
      <Header header={header} />
      <div className="page">
        <div className="content col-md-8">
          {errorContent}
          <div className="buttons">
            <Button
              button={homeButton}
              onClick={() => {
                sessionStorage.clear();
                setToHome(true);
              }}
            />
          </div>
        </div>
      </div>
      <Footer />
    </main>
  );
}

Error.propTypes = {
  page: PropTypes.shape({
    error: PropTypes.object.isRequired,
    header: PropTypes.shape({
      name: PropTypes.string.isRequired
    }).isRequired
  }).isRequired
};
