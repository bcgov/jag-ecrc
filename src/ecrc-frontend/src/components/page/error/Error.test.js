/* eslint-disable no-console */
import React from "react";
import { create } from "react-test-renderer";
import { MemoryRouter, Router } from "react-router-dom";
import { render, fireEvent, getByText, wait } from "@testing-library/react";
import { createMemoryHistory } from "history";

import Error from "./Error";

describe("Error Component", () => {
  const header = {
    name: "Criminal Record Check"
  };

  const error = { status: 403, message: "Unauthorized" };

  const page = {
    header,
    error
  };

  test("Matches the snapshot", () => {
    const ErrorPage = create(
      <MemoryRouter>
        <Error page={page} />
      </MemoryRouter>
    );
    expect(ErrorPage.toJSON()).toMatchSnapshot();
  });

  test("Validate home button click takes user home", async () => {
    const history = createMemoryHistory();
    const { container } = render(
      <Router history={history}>
        <Error page={page} />
      </Router>
    );

    fireEvent.click(getByText(container, "Home"));

    await wait(() => {
      expect(history.location.pathname).toEqual("/");
      expect(sessionStorage.length).toEqual(0);
    });
  });

  test("Validate service unavailable case", async () => {
    console.error = jest.fn();

    const newPage = {
      header,
      error: null
    };

    const history = createMemoryHistory();
    const { container } = render(
      <Router history={history}>
        <Error page={newPage} />
      </Router>
    );

    await expect(() => {
      getByText(container, "wrongtext");
    }).toThrow();

    expect(
      getByText(container, "Service is currently unavailable")
    ).toBeTruthy();
  });

  test("Validate unauthorized 403 case", async () => {
    const history = createMemoryHistory();
    const { container } = render(
      <Router history={history}>
        <Error page={page} />
      </Router>
    );

    await expect(() => {
      getByText(container, "wrongtext");
    }).toThrow();

    expect(getByText(container, "Unauthorized entry")).toBeTruthy();
  });

  test("Validate login failed case", async () => {
    const newError = { status: 403, message: "BCSC login failed" };

    const newPage = {
      header,
      error: newError
    };

    const history = createMemoryHistory();
    const { container } = render(
      <Router history={history}>
        <Error page={newPage} />
      </Router>
    );

    await expect(() => {
      getByText(container, "wrongtext");
    }).toThrow();

    expect(getByText(container, "Login failed")).toBeTruthy();
  });

  test("Validate session expired case", async () => {
    const updatedError = { ...error, status: 590 };

    const newPage = {
      header,
      error: updatedError
    };

    const history = createMemoryHistory();
    const { container } = render(
      <Router history={history}>
        <Error page={newPage} />
      </Router>
    );

    await expect(() => {
      getByText(container, "wrongtext");
    }).toThrow();

    expect(getByText(container, "Your session has expired")).toBeTruthy();
  });

  test("Validate unexpected error case", async () => {
    const updatedError = { ...error, status: 204 };

    const newPage = {
      header,
      error: updatedError
    };

    const history = createMemoryHistory();
    const { container } = render(
      <Router history={history}>
        <Error page={newPage} />
      </Router>
    );

    await expect(() => {
      getByText(container, "wrongtext");
    }).toThrow();

    expect(getByText(container, "An unknown error has occurred")).toBeTruthy();
  });
});
