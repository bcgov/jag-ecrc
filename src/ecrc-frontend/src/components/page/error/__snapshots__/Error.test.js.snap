// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Error Component Matches the snapshot 1`] = `
<main>
  <header>
    <nav
      className="navbar navbar-expand-lg navbar-dark"
    >
      <div
        className="container"
      >
        <a
          className="navbar-brand"
          href="/"
        >
          <img
            alt="B.C. Government Logo"
            className="img-fluid d-none d-md-block"
            height="44"
            src="/images/bcid-logo-rev-en.svg"
            width="181"
          />
          <img
            alt="B.C. Government Logo"
            className="img-fluid d-md-none"
            height="44"
            src="/images/bcid-symbol-rev.svg"
            width="64"
          />
        </a>
        <div>
          <a
            className="navbar-brand nav-item nav-link"
            href="/"
          >
            Criminal Record Check
          </a>
        </div>
        <button
          aria-controls="navbarNavAltMarkup"
          aria-expanded="false"
          aria-label="Toggle navigation"
          className="navbar-toggler"
          data-target="#navbarNavAltMarkup"
          data-toggle="collapse"
          type="button"
        >
          <span
            className="navbar-toggler-icon"
          />
        </button>
        <div
          className="collapse navbar-collapse"
          id="navbarNavAltMarkup"
        />
      </div>
    </nav>
  </header>
  <div
    className="page"
  >
    <div
      className="content col-md-8"
    >
      <div>
        <h1>
          An unknown error has occurred.
        </h1>
        <br />
        <p>
          The error description is below. If this error persists, please try again later.
          <br />
          <br />
          403
        </p>
      </div>
      <div
        className="buttons"
      >
        <button
          className="btn ecrc_go_btn btn"
          onClick={[Function]}
          type="submit"
        >
          Home
        </button>
      </div>
    </div>
  </div>
  <footer
    className="footer"
  >
    <nav
      className="navbar navbar-expand-sm navbar-dark"
    >
      <div
        className="container"
      >
        <ul
          className="navbar-nav"
        >
          <li
            className="nav-item"
          >
            <a
              className="nav-link"
              href="https://www2.gov.bc.ca/gov/content/home/disclaimer"
              rel="noopener noreferrer"
              target="_blank"
            >
              Disclaimer
            </a>
          </li>
          <li
            className="nav-item"
          >
            <a
              className="nav-link"
              href="https://www2.gov.bc.ca/gov/content/home/privacy"
              rel="noopener noreferrer"
              target="_blank"
            >
              Privacy
            </a>
          </li>
          <li
            className="nav-item"
          >
            <a
              className="nav-link"
              href="https://www2.gov.bc.ca/gov/content/home/accessibility"
              rel="noopener noreferrer"
              target="_blank"
            >
              Accessibility
            </a>
          </li>
          <li
            className="nav-item"
          >
            <a
              className="nav-link"
              href="https://www2.gov.bc.ca/gov/content/home/copyright"
              rel="noopener noreferrer"
              target="_blank"
            >
              Copyright
            </a>
          </li>
          <li
            className="nav-item"
          >
            <a
              className="nav-link"
              href="https://www2.gov.bc.ca/gov/content/home/contact-us"
              rel="noopener noreferrer"
              target="_blank"
            >
              Contact Us
            </a>
          </li>
        </ul>
      </div>
    </nav>
  </footer>
</main>
`;
